### ISSUE TITLE
Two releases with a same named package fail

### ISSUE THREAD

**Issue by pc-jedi** [cloudfoundry/bosh#1940]
I have two Bosh-releases which have a `jq` package.
If I deploy one job `elastic-stack/elasticsearch which depends on `elastic-stack/jq` and another job `telegraf/telegraf` which depends on `telegraf/jq` on the same instance group I get the follwoing error

```
Package name collision detected in instance group 'elasticsearch_master': job 'elastic-stack/elasticsearch' depends on package 'elastic-stack/jq', job 'telegraf/telegraf' depends on 'telegraf/jq'. BOSH cannot currently collocate two packages with identical names from separate releases.
```

According to the release notes it should work:
https://github.com/cloudfoundry/bosh/releases/tag/v264.2.0

**Comment by voelzmo** [comment]
Hey Christian,

if I'm not mistaken, it depends on the fingerprints of the packages being the same. This means, if your content in the two packages is the exact same, up to the last bit, this should totally work.

If, however, you do have a difference between the two, this fails with the error you're describing above.

Hope that helps!

**Comment by pc-jedi** [comment]
Hey Marco,
Yeah that helps, but what if both jobs depend on different versions?
And will there be a solution for package re-use in the future? Because I personally hate duplicating code/blobs/etc.

**Comment by voelzmo** [comment]
Well, the solution for *not* duplicating this is to re-use the same package ;)

IIRC, there should be a story which allows packages of the same name from different releases even if fingerprints aren't the same. Does that sound like a story in your backlog @cppforlife?

**Comment by drnic** [comment]
Checkout:
* https://bosh.io/docs/package-vendoring/
* https://www.cloudfoundry.org/blog/build-better-bosh-releases-faster-with-language-packs/

**Comment by cppforlife** [comment]
> IIRC, there should be a story which allows packages of the same name from different releases even if fingerprints aren't the same. Does that sound like a story in your backlog @cppforlife?

we do want to support such configuration, but it's not that high in priority at the moment since most of the time renaming does the job.

**Comment by cppforlife** [comment]
(also, somewhat invasive problem to solve unfortunately).

**Comment by cppforlife** [comment]
> According to the release notes it should work

this would only work when fingerprints and their dependencies match perfectly. that could be achieved via vendor-package cmd.

**Comment by mkuratczyk** [comment]
We just run into this issue as well. Our release contained `golang` package and everything worked fine in our test environments. However, it was failing in another env. It turned out an add-on which also contained `golang` package was installed in that env and on the deployed VM the package was actually coming from the add-on. It contained the same Golang version but their `packaging` script was different which led to a different path for `go` binary.

We've just renamed our package to avoid this conflict but this behaviour leads to some very hard to diagnose issues so it would be great to fix that.

**Comment by mcwumbly** [comment]
We've run into this again recently in PAS as we're looking to co-locate the BOSH CLI on VMs that require it using this package: https://github.com/bosh-packages/cf-cli-release

> IIRC, there should be a story which allows packages of the same name from different releases even if fingerprints aren't the same. 

Is there something else we can follow to stay in the loop on any progress here? Or should it be sufficient to just follow this github issue?

**Comment by dpb587-pivotal** [comment]
There's not a story about this sort of feature. Any future work around this will require some community discussion since it will introduce some new release conventions, so this issue is probably the best place to watch for now until a proposal gets going.

**Comment by mkuratczyk** [comment]
Given that it's a breaking change to support multiple packages, could we instead get the collision error in all cases? As I mentioned above - we had two Golang packages with the same Go version and the same package name (one from our tile, one from an add-on we didn't know would be collocated) and our deployment got broken because the folder structure inside `/var/vcap/packages/golang` was different than expected (the other package "won" and its `packaging` script was different). I think BOSH should prevent situations like that when an addon can break the deployment.

**Comment by ljfranklin** [comment]
@dpb587 solving the problem holistically sounds like it will require a lot of work and coordination. But maybe there's a smaller, easier fix related to the shared bosh-package releases, e.g. [golang-release](https://github.com/bosh-packages/golang-release). Right now if release A and release B are co-located and both vendor the golang-1.10-linux package, they'll get an error if one is on a different patch version of golang correct? If all releases moved to the vendor package flow, I'd imagine this would be a super-common occurrence. To sidestep this issue, would it be better to have the vendored package names include the patch version, e.g. golang-1.10.0-linux? Then when one release bumps to a newer patch version they can still be co-located without waiting for the other release to also bump. Please let me know if I'm misunderstanding something about the vendor package implementation though.

**Comment by dpb587-pivotal** [comment]
@mkuratczyk @ljfranklin to tangent slightly... in this `golang` scenario, does this mean one of the jobs has `golang` listed as a job dependency? Typically `golang` is only necessary as a package dependency, which I believe sidesteps the issue since compilation dependencies are not considered for package name overlap. Might be worth double checking job dependencies in those cases.

@ljfranklin yeah, using package names with exact version information would be an alternative. In this case, we have avoided that on `golang` because it has historically been a compile-only dependency, so we stop at the minor version. But for things like the `ruby-release` we did introduce a slightly odd revision suffix (e.g. `ruby-2.4-r4`) to help avoid conflicts in CPI world, among other reasons.

**Comment by mkuratczyk** [comment]
@dpb587 we've solved this problem by renaming our golang package for now and we are thinking about switching to golang-release in the future but just to explain - we use Golang and ginkgo to run smoke tests (RMQ tile). I'm sure we could do it differently but for the time being that's how we execute smoke tests errand and therefore we need Golang as a job dependency, not for compilation.

**Comment by mfine30** [comment]
This is definitely something we do want to get to, but the solution is unfortunately not trivial. I'll keep this issue open to track concerns and discussion, but still probably a way off before we can dedicate time to really digging into it.

**Comment by Jrmcneil** [comment]
Hey @mfine30 any updates on this? Rabbit is having this issue again.
--
@dlresende @ferozjilla
